src/ctap/storage.rs

```diff
@@ -532,145 +416,186 @@ impl PersistentStore {
-                self.store.insert(entry).unwrap();
-            }
-            Some((index, _)) => {
-                self.store.replace(index, entry).unwrap();
-            }
-        }
-        Ok(())
+        Ok(self.store.insert(
+            key::_MIN_PIN_LENGTH_RP_IDS,
+            &_serialize_min_pin_length_rp_ids(min_pin_length_rp_ids)?,
+        )?)
     }
 
+    /// Returns the attestation private key if defined.
     pub fn attestation_private_key(
         &self,
-    ) -> Result<Option<&[u8; key_material::ATTESTATION_PRIVATE_KEY_LENGTH]>, Ctap2StatusCode> {
-        let data = match self.store.find_one(&Key::AttestationPrivateKey) {
-            None => return Ok(None),
-            Some((_, entry)) => entry.data,
-        };
-        if data.len() != key_material::ATTESTATION_PRIVATE_KEY_LENGTH {
-            return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+    ) -> Result<Option<[u8; key_material::ATTESTATION_PRIVATE_KEY_LENGTH]>, Ctap2StatusCode> {
+        match self.store.find(key::ATTESTATION_PRIVATE_KEY)? {
+            None => Ok(None),
+            Some(key) if key.len() == key_material::ATTESTATION_PRIVATE_KEY_LENGTH => {
+                Ok(Some(*array_ref![
+                    key,
+                    0,
+                    key_material::ATTESTATION_PRIVATE_KEY_LENGTH
+                ]))
+            }
+            Some(_) => Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INVALID_PERSISTENT_STORAGE),
         }
-        Ok(Some(array_ref!(
-            data,
-            0,
-            key_material::ATTESTATION_PRIVATE_KEY_LENGTH
-        )))
     }
 
+    /// Sets the attestation private key.
+    ///
+    /// If it is already defined, it is overwritten.
     pub fn set_attestation_private_key(
         &mut self,
         attestation_private_key: &[u8; key_material::ATTESTATION_PRIVATE_KEY_LENGTH],
     ) -> Result<(), Ctap2StatusCode> {
-        let entry = StoreEntry {
-            tag: ATTESTATION_PRIVATE_KEY,
-            data: attestation_private_key,
-            sensitive: true,
-        };
-        match self.store.find_one(&Key::AttestationPrivateKey) {
-            None => self.store.insert(entry)?,
-            _ => return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR),
+        match self.store.find(key::ATTESTATION_PRIVATE_KEY)? {
+            None => Ok(self
+                .store
+                .insert(key::ATTESTATION_PRIVATE_KEY, attestation_private_key)?),
+            Some(_) => Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR),
         }
-        Ok(())
     }
 
+    /// Returns the attestation certificate if defined.
     pub fn attestation_certificate(&self) -> Result<Option<Vec<u8>>, Ctap2StatusCode> {
-        let data = match self.store.find_one(&Key::AttestationCertificate) {
-            None => return Ok(None),
-            Some((_, entry)) => entry.data,
-        };
-        Ok(Some(data.to_vec()))
+        Ok(self.store.find(key::ATTESTATION_CERTIFICATE)?)
```

Comment:
Yes, it's some kind of magic :D The `?` operator doesn't simply return in case of error, it also converts the error to the expected type (using `From`). So in this case (and other places in this file), `Ok(x?)` converts a `Result<T, StoreError>` to `Result<T, Ctap2StatusCode>` using the mapping defined in `impl From<StoreError> for Ctap2StatusCode`.